### ISSUE TITLE
[Feature Request] Steam OTP Support

### ISSUE THREAD

**Issue by igaryhe** [tadfisher/pass-otp#97]
Steam seems like using a 5 character length otp key other than 6 character length digits. KeeTrayTOTP (https://github.com/victor-rds/KeeTrayTOTP) has supported this format, and this may be a quite important feature for pass users as well.

**Comment by tadfisher** [comment]
The key URI format supports a "digits" parameter for setting arbitrary code length. Have you tried that?

**Comment by igaryhe** [comment]
I invested the code and realized that you implement this feature using oathtool? Maybe this is a feature that need to be implemented upstream. Anyway, according to the implementation in KeeTrayTOTP, instead of mod the truncated decimal number with a given number, e.g. 10^6 when digits=6, they use another function to calculate the last part of the TOTP algorithm. Don't know how to explicitly express the method, but you could spend some time read this piece of code: https://github.com/victor-rds/KeeTrayTOTP/blob/ffcce347538dde332cacb24aaba8b5b2909a0fbc/Libraries/TOTPEncoder.cs

**Comment by tadfisher** [comment]
Okay, oathtool won't support Steam Guard codes.

First, it only allows a `digits` value in 6, 7, or 8.

Second, we don't get the full token generated by the hash algorithm prior to the mod operations which generate the code.

So either oathtool needs to support both of those things, or we need to find a different OTP tool which does, preferably one that is suitably cross-platform and available.

**Comment by igaryhe** [comment]
Yeah I'm trying to patch authtool to make it generate correct Steam Guard code. However I stuck at compiling the code...

**Comment by tadfisher** [comment]
Perusing the project's list of merge requests, you're going to have an uphill battle getting any patch merged.

This might be a good reason to rewrite this package in a native language, or at least try and popularize a better tool. I don't generally like pass extensions that require a bunch of dependencies, but something like Rust or Go would be perfect for static binaries while still being easily buildable by distributions.

**Comment by igaryhe** [comment]
Found some Rust projects which have been dealing with OTP functions. I'm learning Rust recently, maybe I can create a tool which is easier for you to integrate. I'm not familiar with bash code, could you enumerate which commands of oathtool are being used in your project?

**Comment by igaryhe** [comment]
I wrote a small tool, which implements the basic functions, like HOTP and TOTP (only HMAC-SHA1 at the moment): https://github.com/igaryhe/rotp

**Comment by tadfisher** [comment]
That's pretty cool. If maybe there was a Key URI parsing crate as well, we
could possibly remove a bunch of problematic code from pass-otp.

I am wary about packaging, though. oathtool seems to be packaged for most
Unices and I am not sure what sort of security review some distributions
need for a new OTP tool, e.g. Debian. I wouldn't want to blindside them
with a new unpackaged dependency either.

Do you have thoughts on the above?

**Comment by igaryhe** [comment]
I'm going to do some refinement and add URI parsing function. Maybe then try to upload the tool to some package manager.

**Comment by lanodan** [comment]
As maintainer of the `pass-otp` ebuild for gentoo, rust seems to be something quite new so packaging this dependency might be quite hard.

As myself: It looks like your program code is quite simple compared to the large size of oathtool(1) from oath-toolkit (which also looks unmaintained), so probably easier to audit and improve, I’m often quite skeptical of rewrites but it looks worth it here.

**Comment by NilsIrl** [comment]
> This might be a good reason to rewrite this package in a native language

This was also mentioned by Jason Donenfeld on the mailing list as something that should be done for `pass`. (Though nothing seems to have been done so far)

**Comment by baierjan** [comment]
I have an alternative solutions for this problem. It seems that the implementation of HOTP/TOTP is pretty simple and straight-forward. So I made a drop-in replacement for oathtool (it should correctly takes all options which are currently in use) using  [Pass::OTP](https://github.com/baierjan/Pass-OTP-perl).

Then I took it one step further and I enhanced the wrapper to actually understand the `otpauth://` links, so it might be possible to remove some code. It also understand the issuer and if it detect Steam, it will use alternative char set, so it produces correct Steam Guard code.

It is written in Perl, so the packaging and compatibility should be OK.
